internal/proxy/proxy.go

```diff
@@ -311,22 +242,92 @@ type socketMount struct {
 	listener net.Listener
 }
 
-// listen causes a socketMount to create a Listener at the specified network address.
-func (s *socketMount) listen(ctx context.Context, network string, address string) (net.Addr, error) {
+func newSocketMount(ctx context.Context, conf *Config, pc *portConfig, inst InstanceConnConfig, version string) (*socketMount, error) {
+	var (
+		// network is one of "tcp" or "unix"
+		network string
+		// address is either a TCP host port, or a Unix socket
+		address string
+	)
+	// IF
+	//   a global Unix socket directory is NOT set AND
+	//   an instance-level Unix socket is NOT set
+	//   (e.g.,  I didn't set a Unix socket globally or for this instance)
+	// OR
+	//   an instance-level TCP address or port IS set
+	//   (e.g., I'm overriding any global settings to use TCP for this
+	//   instance)
+	// use a TCP listener.
+	// Otherwise, use a Unix socket.
+	if (conf.UnixSocket == "" && inst.UnixSocket == "") ||
+		(inst.Addr != "" || inst.Port != 0) {
+		network = "tcp"
+
+		a := conf.Addr
+		if inst.Addr != "" {
+			a = inst.Addr
+		}
+
+		var np int
+		switch {
+		case inst.Port != 0:
+			np = inst.Port
+		case conf.Port != 0:
+			np = pc.nextPort()
+		default:
+			np = pc.nextDBPort(version)
+		}
+
+		address = net.JoinHostPort(a, fmt.Sprint(np))
+	} else {
+		network = "unix"
+
+		dir := conf.UnixSocket
+		if dir == "" {
+			dir = inst.UnixSocket
+		}
+		if _, err := os.Stat(dir); err != nil {
+			return nil, err
+		}
+		address = UnixAddress(dir, inst.Name)
+		// When setting up a listener for Postgres, create address as a
+		// directory, and use the Postgres-specific socket name
+		// .s.PGSQL.5432.
+		if strings.HasPrefix(version, "POSTGRES") {
+			// Make the directory only if it hasn't already been created.
+			if _, err := os.Stat(address); err != nil {
+				if err = os.Mkdir(address, 0777); err != nil {
+					return nil, err
+				}
+			}
+			address = UnixAddress(address, ".s.PGSQL.5432")
+		}
+	}
+
 	lc := net.ListenConfig{KeepAlive: 30 * time.Second}
-	l, err := lc.Listen(ctx, network, address)
+	ln, err := lc.Listen(ctx, network, address)
 	if err != nil {
 		return nil, err
 	}
-	s.listener = l
-	return s.listener.Addr(), nil
+	var opts []cloudsqlconn.DialOption
+	if inst.IAMAuthN != nil {
+		opts = append(opts, cloudsqlconn.WithDialIAMAuthN(*inst.IAMAuthN))
+	}
+	m := &socketMount{inst: inst.Name, dialOpts: opts, listener: ln}
+	return m, nil
+}
+
+func (s *socketMount) Addr() net.Addr {
+	return s.listener.Addr()
+}
+
+func (s *socketMount) Accept() (net.Conn, error) {
+	return s.listener.Accept()
 }
 
 // close stops the mount from listening for any more connections
-func (s *socketMount) close() error {
-	err := s.listener.Close()
-	s.listener = nil
-	return err
+func (s *socketMount) Close() error {
+	return s.listener.Close()
```

Comment:
We used to make the listener null. Why not make it null again?

Reply:
That was the root of the race condition. Mutating listener across goroutines.